Entities bug-fixes in place management

benbosman opened the pull request:
We've noticed a few use cases where the place of a new relationship is not defined correctly. These problems have been solved and JUnit tests have been created to automatically verify the functionality.

The previously existing tests didn't cause any problems because:
* All tests added the relationship as the last element
* There were no tests to change the place of a relationship
* The item.update was always called somewhere implicitly to fix the places

All of the previous tests still succeed, and the additional tests succeed as well.
Inline comments have been included to explain the more complicated reasons for the approach in determining the place
The JavaDocs have been updated as well

Raf-atmire committed:
[Task 65568] initial fixes and test fixes

Raf-atmire committed:
[Task 65568] implemented tests and fixed checkstyle

Raf-atmire committed:
Merge remote-tracking branch 'dspace/master' into w2p-65568_bugs-place-management-creating-relationships

Conflicts:
	dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/RelationshipRestRepository.java

Raf-atmire committed:
Fixes after master merge

Raf-atmire committed:
[Task 65568] updated javadoc and comments

Raf-atmire committed:
[Task 65681] updated RelationshipSeviceImplTes

Raf-atmire committed:
Merge remote-tracking branch 'dspace/master' into w2p-65568_bugs-place-management-creating-relationships

Conflicts:
	dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java

Raf-atmire committed:
Fix after merge

benbosman committed:
Travis build trigger

paulo-graca commented on dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java:
```diff
@@ -778,8 +778,8 @@ private void addRelationship(Context c, Item item, String typeName, String value
         // Create the relationship
-        int leftPlace = relationshipService.findLeftPlaceByLeftItem(c, leftItem) + 1;
-        int rightPlace = relationshipService.findRightPlaceByRightItem(c, rightItem) + 1;
+        int leftPlace = relationshipService.findNextLeftPlaceByLeftItem(c, leftItem);
```
The first time I executed an CSV import, I got an error. On the second try, without any changes, the record was registered. But I noticed a duplication of isAuthorOfPublication on the Metadatafield with the relationship.

`metadata_field_id: 134 (relation.isAuthorOfPublication)`
![2019-10-29 09_52_48-Valentina Studio](https://user-images.githubusercontent.com/877488/67757563-f871e280-fa33-11e9-981d-dcb737c5388a.png)

Should a specific relation.isAuthorOfPublication should be present in this table?

benbosman replied:
Hi @paulo-graca,

I was able to reproduce the problem, but also on master. It seems you've found a bug in one of the CSV branches already merged in master.
Can you move this to a separate ticket so it's not lost?

I was able to reproduce this with the CSV https://www.dropbox.com/s/vvugb1tuknjhp9a/test-records-2.csv?dl=1 which created relation.isAuthorOfPublication with value 0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f twice using master

MarieVerdonck committed:
65942: Merge branch 'master' into w2p-65568_bugs-place-management-creating-relationships

# Conflicts:
#	dspace-api/src/main/java/org/dspace/content/dao/RelationshipDAO.java

MarieVerdonck committed:
65942: Test fix since implementation changed as well

tdonohue commented on dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java:
```diff
@@ -94,18 +98,32 @@ public Relationship create(Context context, Relationship relationship) throws SQ
+        }
+        Item rightItem = relationship.getRightItem();
+        if (relationship.getRightPlace() == -1) {
+            relationship.setRightPlace(Integer.MAX_VALUE);
```
This method's logic  is increasingly confusing.  Can you better explain why we have to temporarily set the place to `MAX_VALUE`?  Why not just set it to the length of `leftRelationships` or `rightRelationships` on the next lines?

benbosman replied:
The place -1 implies it's not set and should be appended at the end.
The length of the arrays is also not necessarily the correct place, e.g. if there are plain text authors present

A comment has been included at https://github.com/DSpace/DSpace/pull/2551/files/e56e71a90338000666a2ce012c3849fc56d0c639#diff-2e89fe18482e7ac43a1a4e901a8105b7R104 to explain this

tdonohue commented on dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java:
```diff
@@ -94,18 +98,32 @@ public Relationship create(Context context, Relationship relationship) throws SQ
+            leftRelationships.remove(relationship);
+        }
+        if (rightRelationships.contains(relationship)) {
+            rightRelationships.remove(relationship);
```
These if statements should be preceded by a comment to describe why we are removing the relationship from the list.  I don't understand the logic here.  Could we add some Unit Tests to prove out this is working as expected?

Raf-atmire replied:
These relationships here aren't removed from the actual list, they're only removed in this temporary list so that we can perform a proper place calculation further on in this method. I do agree that it was a bit unclear, so I'll add a comment stating this. This is particularly useful incase this method is called for a delete use case, so that we can update the relationships' places properly without the relationship given interfering with the place calculation

benbosman replied:
This has been clarified using inline comments

tdonohue commented on dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java:
```diff
@@ -0,0 +1,374 @@
+    }
+
+    @Test
+    public void addMetadataAndRelationshipTest() throws Exception {
```
Either break up this method into separate tests, or please add inline comments.  This method is hard to follow as you create new content midway through and then run more tests.  It's difficult for me to see what exactly you are testing for in each test without understanding all 50 lines of this method.

benbosman replied:
Inline comments have been included

tdonohue commented on dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java:
```diff
@@ -0,0 +1,374 @@
+    }
+
+    @Test
+    public void AddMetadataAndRelationshipWithSpecificPlaceTest() throws Exception {
```
Again, method is too long, and you add new content at several stages.  Please break this up into several smaller tests (with more descriptive names)...or add descriptive inline comments

benbosman replied:
Inline comments have been included

tdonohue commented on dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java:
```diff
@@ -0,0 +1,374 @@
+
+
+    @Test
+    public void AddAndRemoveMetadataAndRelationshipsTest() throws Exception {
```
Same as above.  Too large and no comments

benbosman replied:
Inline comments have been included

tdonohue commented on dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java:
```diff
@@ -0,0 +1,374 @@
+        assertThat(list.size(), equalTo(2));
+
+        // Check that these places are still intact after the deletion as the place doesn't get updated until an
+        // item.update has been called
```
This is a better inline comment.  But, where do we call `item.update` below to ensure the place *does* get updated after the deletion?

benbosman replied:
it happens implicitly in https://github.com/DSpace/DSpace/pull/2551/files/e56e71a90338000666a2ce012c3849fc56d0c639#diff-56b0380ff7b66daeff6012e659df5863R246 but it can be duplicated explicitly as well if that makes it easier to understand

tdonohue reviewed (commented):
@benbosman : I gave this a quick review today. I'm confused by several of the code changes here (see below).  As the new test methods are very large an uncommented, it's difficult to follow them as well.  I'd appreciate either inline comments in large test methods, or find a way to break them up into smaller these methods (where plausible).

benbosman commented on dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java:
```diff
@@ -86,9 +86,9 @@ public int findRightPlaceByRightItem(Context context, Item item) throws SQLExcep
         list.sort((o1, o2) -> o2.getLeftPlace() - o1.getLeftPlace());
         if (!list.isEmpty()) {
-            return list.get(0).getLeftPlace();
+            return list.get(0).getLeftPlace() + 1;
```
we'll check this out

benbosman replied:
This has been solved.
The `testGetNextRightPlace` method has been created to verify this works correctly

Raf-atmire committed:
Inprogress evaluating feedback with regards to comments in RelationshipServiceImpl and the place tests

Raf-atmire committed:
Fixed issue in findNextRightPlaceByRightItem and added a test for it. Also wrote additional docs

paulo-graca reviewed (approved):
The changes look ok to me. And didn't found any issues (besides the CSV import) within my tests. My thanks to @benbosman and @Raf-atmire for this contribution!

tdonohue commented on dspace-api/src/test/java/org/dspace/content/RelationshipMetadataServiceTest.java:
```diff
@@ -138,5 +138,41 @@ public void testGetRelationshipMetadata() throws Exception {
+    }
+
+    @Test
+    public void testGetNextRightPlace() throws Exception {
```
Could we add a corresponding test for `findNextLeftPlaceByLeftItem`?  I just noticed it seems to be missing.

tdonohue reviewed (commented):
@benbosman : This is looking good overall now.  Just one (very minor) missing test...see below.  Once that's ready, I think this can be merged.

Raf-atmire committed:
[Task 65568] added a testGetNextLeftPlace test in the RelationshipMetadataServiceTest

tdonohue reviewed (approved):
👍 Thanks @benbosman and @Raf-atmire.  Looks good to me!
